File: tools/peanut-review/peanut_review/web/assets/queue.js
// Local review queue. Review tasks are handed to the user's driver agent.
(() => {
  "use strict";
  const base = window.PR_BASE_URL || "";
  const byId = (id) => document.getElementById(id);
  let data = { items: [], accounts: [] };
  let loading = false;
  let submitting = false;
  const themes = ["system", "dark-plus", "light"];
  let theme = "system";
  try { theme = localStorage.getItem("pr.theme") || "system"; } catch { /* optional */ }
  function applyTheme() {
    if (theme === "system") document.documentElement.removeAttribute("data-theme");
    else document.documentElement.dataset.theme = theme;
    byId("theme-toggle").textContent = `theme: ${theme}`;
  }
  byId("theme-toggle").addEventListener("click", () => {
    theme = themes[(themes.indexOf(theme) + 1) % themes.length];
    try { localStorage.setItem("pr.theme", theme); } catch { /* optional */ }
    applyTheme();
  });
  applyTheme();

  function node(tag, text, className = "") {
    const element = document.createElement(tag);
    if (text != null) element.textContent = text;
    if (className) element.className = className;
    return element;
  }
  function relativeTime(value) {
    if (!value) return "Never checked";
    const seconds = Math.max(0, (Date.now() - Date.parse(value)) / 1000);
    if (seconds < 60) return "Checked just now";
    if (seconds < 3600) return `Checked ${Math.floor(seconds / 60)}m ago`;
    return `Checked ${Math.floor(seconds / 3600)}h ago`;
  }
  function error(message) {
    byId("queue-error").textContent = message;
    byId("queue-error").hidden = !message;
  }
  async function copyTask(item) {
    try {
      await navigator.clipboard.writeText(item.driver_task);
      byId("queue-copy-status").textContent = "Task copied. Paste it into your driver conversation.";
    } catch {
      const text = byId("queue-task-text");
      text.value = item.driver_task;
      byId("queue-task-dialog").showModal();
      text.focus();
      text.select();
    }
  }
  function visible(item) {
    const query = byId("queue-search").value.trim().toLowerCase();
    const account = byId("queue-account").value;
    if (account && item.account_key !== account) return false;
    const requestType = byId("queue-request-type").value;
    if (requestType && requestType !== (item.request_kind || "following")) return false;
    if (query && !`${item.repo} ${item.number} ${item.title || ""} ${item.author || ""}`.toLowerCase().includes(query)) return false;
    const filter = byId("queue-filter").value;
    if (filter !== "all" && item.state && item.state !== "open") return false;
    if (filter === "requested") return item.requested;
    if (filter === "stale") return item.freshness === "stale";
    if (filter === "running") return item.session_progress?.status === "running";
    if (filter === "attention") return item.requested || item.freshness !== "current" || item.session_progress?.status === "running" || item.session_progress?.status === "failed";
    return true;
  }
  function render() {
    const accounts = byId("queue-accounts");
    accounts.replaceChildren();
    const selected = byId("queue-account").value;
    byId("queue-account").replaceChildren(new Option("All accounts", ""));
    for (const account of data.accounts) {
      byId("queue-account").add(new Option(account.label, account.key));
      const card = node("div", null, `queue-account-card${account.error ? " account-error" : ""}`);
      card.append(node("strong", account.label), node("span", `@${account.login} · ${account.hostname}`, "meta"));
      card.append(node("span", account.error || (data.refreshing ? "Checking GitHub…" : relativeTime(account.checked_at)), account.error ? "error" : "meta"));
      accounts.append(card);
    }
    byId("queue-account").value = selected;
    const rows = data.items.filter(visible);
    byId("queue-count").textContent = `${rows.length} of ${data.items.length} PRs`;
    byId("queue-refresh").disabled = submitting || data.refreshing;
    byId("queue-refresh").textContent = data.refreshing ? "Refreshing…" : "Refresh queue";
    const body = byId("queue-rows");
    body.replaceChildren();
    for (const item of rows) {
      const row = node("tr", null, "session-row");
      const pr = node("td", null, "queue-pr");
      const link = node("a", item.title || `${item.repo} #${item.number}`);
      // Construct URLs from validated identity fields, never arbitrary PR text.
      link.href = `https://${item.account.hostname}/${item.repo}/pull/${item.number}`;
      link.target = "_blank";
      link.rel = "noopener noreferrer";
      pr.append(link, node("div", `${item.repo} #${item.number}${item.author ? ` · ${item.author}` : ""}${item.draft ? " · Draft" : ""}${item.state && item.state !== "open" ? ` · ${item.state}` : ""}`, "sub"));
      const identity = node("td");
      const account = data.accounts.find((a) => a.key === item.account_key);
      const requestLabel = { direct: "Direct request", team: "Team request" }[item.request_kind] || "Following";
      identity.append(node("div", account?.label || item.account.login), node("div", requestLabel, "sub"));
      if (item.requested) identity.append(node("div", "Review requested", "sub"));
      else if (item.request_kind) identity.append(node("div", "No pending request", "sub"));
      if (item.request_kind_error) identity.append(node("div", item.request_kind_error, "queue-detail error"));
      const review = node("td");
      const progress = item.session_progress?.status || "pending";
      review.append(node("span", item.session_progress?.label || "Not started", `badge review-progress progress-${progress}`));
      const freshness = node("td");
      const freshLabels = { current: "Up to date", stale: "Stale", unknown: "Unknown", unreviewed: "Not reviewed" };
      freshness.append(node("span", freshLabels[item.freshness], `queue-freshness freshness-${item.freshness}`));
      const detail = item.completed_snapshot ? `Reviewed ${item.completed_snapshot.head_sha.slice(0, 8)} · latest ${(item.head_sha || "").slice(0, 8)}` : item.head_sha ? `Latest ${item.head_sha.slice(0, 8)}` : "";
      freshness.append(node("div", detail, "sub mono"), node("div", relativeTime(item.checked_at), "sub"));
      if (item.error) freshness.append(node("div", item.error, "queue-detail error"));
      const actions = node("td", null, "queue-actions");
      if (item.session_id) {
        const open = node("a", "Open review", "queue-open");
        open.href = `${base}/${encodeURIComponent(item.session_id)}`;
        actions.append(open);
      }
      if (item.state === "open") {
        const copy = node("button", item.session_id ? "Copy re-review task" : "Copy review task");
        copy.type = "button";
        copy.disabled = !item.driver_task;
        copy.addEventListener("click", () => copyTask(item));
        actions.append(copy);
      }
      row.append(pr, identity, review, freshness, actions);
      body.append(row);
    }
    byId("queue-empty").hidden = rows.length > 0;
    byId("queue-empty").textContent = data.refreshing && !data.items.length ? "Checking your accounts for review requests…" : data.accounts.some((a) => a.error) && !data.items.length ? "Could not load every account. See account details above and refresh to retry." : "No PRs match this view.";
    document.querySelector(".queue-table").hidden = rows.length === 0;
  }
  async function refresh() {
    if (loading || !window.PR_QUEUE_ENABLED) return;
    loading = true;
    try {
      const response = await fetch(`${base}/api/queue`, { cache: "no-store" });
      if (!response.ok) throw new Error(`Could not load queue (${response.status})`);
      data = await response.json();
      render();
    } catch (err) { error(err.message); }
    finally { loading = false; }
  }
  async function action(name, body = {}) {
    submitting = true;
    error("");
    render();
    try {
      const response = await fetch(`${base}/api/queue/${name}`, {
        method: "POST", headers: { "Content-Type": "application/json", "X-Peanut-Queue-Token": window.PR_QUEUE_TOKEN },
        body: JSON.stringify(body),
      });
      const result = await response.json();
      if (!response.ok) throw new Error(result.error || `Action failed (${response.status})`);
    } catch (err) { error(err.message); }
    finally { submitting = false; await refresh(); render(); }
  }
  byId("queue-refresh").addEventListener("click", () => action("refresh"));
  byId("queue-task-close").addEventListener("click", () => byId("queue-task-dialog").close());
  byId("queue-search").addEventListener("input", render);
  byId("queue-account").addEventListener("change", render);
  byId("queue-request-type").addEventListener("change", render);
  byId("queue-filter").addEventListener("change", render);
  if (window.PR_QUEUE_ENABLED) {
    refresh();
    setInterval(refresh, 3000);
  } else byId("queue-refresh").disabled = true;
})();

File: tools/peanut-review/peanut_review/web/queue.py
"""Review queue page; data and actions are served by the existing HTTP app."""
from __future__ import annotations

import html
import json

from .render import FAVICON_HREF, THEME_BOOTSTRAP, THEME_TOGGLE_BUTTON, _asset_url


def render_queue(base_url: str, token: str, enabled: bool) -> str:
    base = html.escape(base_url, quote=True)
    setup = "" if enabled else '<p class="queue-notice">Enable the queue with <code>peanut-review serve --queue-config /path/to/queue.json</code>. See the review queue setup in the README.</p>'
    return f'''<!doctype html>
<html lang="en"><head><meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>peanut-review — review queue</title>
<link rel="icon" href="{FAVICON_HREF}">
<link rel="stylesheet" href="{html.escape(_asset_url(base_url, "style.css"), quote=True)}">
{THEME_BOOTSTRAP}</head>
<body class="index queue-page">
<header>
<h1><a href="{base}/">🥜 peanut-review</a></h1>
<nav class="app-nav" aria-label="Dashboard"><a href="{base}/queue" aria-current="page">Review queue</a><a href="{base}/">Sessions</a></nav>
<span class="spacer"></span>{THEME_TOGGLE_BUTTON}
<button id="queue-refresh" type="button">Refresh queue</button>
</header>
<main class="index-main">
<div class="queue-heading"><div><h2>Your review queue</h2><p class="meta">Review requests and tracked PRs across your accounts.</p></div><span id="queue-count" class="meta" aria-live="polite"></span></div>
<p class="meta">Copy a review task into your driver conversation to prepare the checkout and run the review.</p>
{setup}
<div id="queue-accounts" class="queue-accounts"></div>
<div class="queue-toolbar">
<input id="queue-search" type="search" placeholder="Search PRs, repositories, authors…" aria-label="Search review queue">
<select id="queue-account" aria-label="Filter by account"><option value="">All accounts</option></select>
<select id="queue-request-type" aria-label="Filter by request type">
<option value="">All request types</option><option value="team">Team request</option><option value="direct" selected>Direct request</option>
</select>
<select id="queue-filter" aria-label="Filter review queue">
<option value="attention">Needs attention</option><option value="open">All open PRs</option><option value="requested">Review requested</option><option value="stale">Stale reviews</option><option value="running">Running reviews</option><option value="all">Include closed PRs</option>
</select></div>
<p id="queue-error" class="queue-notice error" role="alert" hidden></p>
<p id="queue-copy-status" class="meta" role="status"></p>
<div class="queue-table-wrap"><table class="sessions queue-table">
<thead><tr><th>Pull request</th><th>Account / request</th><th>Review</th><th>Freshness</th><th>Actions</th></tr></thead>
<tbody id="queue-rows"></tbody></table></div>
<div id="queue-empty" class="empty">{"Loading your review queue…" if enabled else "Queue is not configured."}</div>
</main>
<dialog id="queue-task-dialog" aria-labelledby="queue-task-title">
<h2 id="queue-task-title">Copy task to your driver</h2>
<p>Automatic copying is unavailable. Copy the selected text and paste it into your driver conversation.</p>
<textarea id="queue-task-text" aria-label="Review task" rows="14" readonly></textarea>
<button id="queue-task-close" type="button">Close</button>
</dialog>
<script>window.PR_BASE_URL = {json.dumps(base_url).replace('<', chr(92) + 'u003c')}; window.PR_QUEUE_TOKEN = {json.dumps(token)}; window.PR_QUEUE_ENABLED = {json.dumps(enabled)};</script>
<script src="{html.escape(_asset_url(base_url, 'queue.js'), quote=True)}"></script>
</body></html>'''
